src/common/io/io.cpp

```diff
@@ -573,6 +588,46 @@ error::Error FileReader::Rewind() {
 	return error::NoError;
 }
 
+ExpectedSize BufferedReader::Read(vector<uint8_t>::iterator start, vector<uint8_t>::iterator end) {
+	if (rewind_done_ && !rewind_consumed_) {
+		// Read from the buffer
+		auto ex_bytes_read = buffer_reader_.Read(start, end);
+		if (!ex_bytes_read) {
+			return ex_bytes_read;
+		}
+
+		Vsize bytes_read_buffer = ex_bytes_read.value();
+		if (bytes_read_buffer > 0) {
+			return bytes_read_buffer;
+		}
+		// When EOF, continue with reading from the wrapped reader
+		rewind_consumed_ = true;
+	}
+
+	// Read from the wrapped reader and save copy into the buffer
+	auto bytes_read = wrapped_reader_.Read(start, end);
+	if (!bytes_read) {
+		return bytes_read;
+	}
+	if (!stop_done_) {
+		buffer_.insert(buffer_.end(), start, end);
+	}
+	if (rewind_consumed_ && stop_done_) {
```

Comment:
Micro-optimization/readability: You can replace with `} else if (rewind_consumed_) {` here.